[Backend] NodeJS - week 2 (#199)
HackYourFuture-CPH/program: courses/backend/node/week2/session-plan.md

# Session plan

## Session outline

- [Database interaction with Knex](#database-interaction-with-knex)
- [Snippets API continued](#snippets-api-continued)
- [Advanced Postman use cases](#advanced-postman)

## Database interaction with Knex

Trainees have used Knex before. In foundation, they used it with the .raw() command to execute SQL easily. And they also used it last week when learning about Express.

This part of the module should explain Knex in a lot more technical detail. The following implementation is how we expect to see trainees using database interaction from this point forward.

> 1. Explain what Knex is in more detail, and why we are using it here

Author: Considering that mentors and trainees both have access to this file AND most commonly mentors will use this files during the class on the screen, I would suggest rephrasing these instructions to less directed:

```suggestion
1. What is Knex, and why we need it
```

   - Knex is SQL query builder for Node, which helps us write databases queries more cleanly, without raw SQL.
   - It supports multiple databases (e.g. SQLite, Postgres) with the same API.
   - It helps avoid security issues like SQL injection by handling things safely for us.
   - It provides advanced features such as migrations.

> 2. Compare using .raw() to the Query Builder methods

Author: ```suggestion
2. Comparison: `.raw()` method vs. the Query Builder methods
```

   - Using the Query Builder methods like `.select()` help us write safer, more readable DB queries that work even if we change our database type.

### Live coding

Run through the [phonebook example](./session-materials/phonebook/). The code is already written, but feel free to clear some of the functions so you can write them together in the session.

1. Walk through `database.js` and explain the setup
2. Create the database following the instructions in `database.js`
3. Run through `phonebook.js` function by function, pointing to the documentation which explains all the query builder methods.
4. After writing each function, test it by running it via the main function.

## Snippets API continued

Now we can pick up where we left the exercises last week. Help the trainees complete the remaining endpoints:

TODO: Add a section with live coding somewhere below to explain error handling? include that in one/all of the exercises?

1. [POST endpoint exercise](./session-materials/07-post-endpoint.md)
2. [PUT endpoint exercise](./session-materials/08-put-endpoint.md)
3. [DELETE endpoint exercise](./session-materials/09-delete-endpoint.md)

## Advanced Postman

> Postman can be used for quickly testing your APIs, but can also be configured in more advanced ways to support your development workflow. Here, you'll learn about four ways to level up your Postman game.

Reviewer: My understanding is different. It's not exactly strongly held, but for the new program we have been writing session plans assuming the reader is the mentor. Of course there could be useful content in here, and trainees are free to read/reference it.

The reason I have written some content that sounds like it's for trainees is because i thought of sections like this almost like scripts. So i imagine the mentor leaning on this content as inspiration of what to say. Like imagine if i put quotation marks around this paragraph :D 

I totally see now that it reads quite mixed up though, as you say. 

We have had a [short discussion on slack](https://hackyourfuture-cph.slack.com/archives/C070USK1UET/p1747647691745419) about this back in summary, which is how i ended up with this perspective.

For now, I suggest we stick with that. Although, it's open for discussion (as most things are!).

Regarding moving this PR forward specifically, i'm not sure how to make it clearer for now, but i'd be keen to merge it if the content itself is good and aligns well with the learning goals. Once that's good, we can iterate on the wording and written content some more.

Would you be happy to proceed in that way too? 🤔

> Postman can be used for quickly testing your APIs, but can also be configured in more advanced ways to support your development workflow. Here, you'll learn about four ways to level up your Postman game.

Reviewer: Just realised too, that I write exercises from the perspective of the trainee since they are instructions for them to complete. This also adds to the mix of tones in the session plan.

> Postman can be used for quickly testing your APIs, but can also be configured in more advanced ways to support your development workflow. Here, you'll learn about four ways to level up your Postman game.

Author: I think I still don't get it and I'm quite confused because previous modules I wrote had different perspective in mind and received no feedback on it :) But assuming session plan is for the mentor, then mixing perspective in it is also confusing for the mentor I believe 🤔 

Anyway, I agree with this one:

> Regarding moving this PR forward specifically, i'm not sure how to make it clearer for now, but i'd be keen to merge it if the content itself is good and aligns well with the learning goals. Once that's good, we can iterate on the wording and written content some more.

Could be one of the subtasks to clear it up :)

> Note: These exericses could be demod by the mentor, or left to the trainees to do on their own machines. Or a mix of both, whatever feels natural.

Reviewer: That's fair. They are called exercises, which makes it clear the trainees need to complete them in my view. It's up to the mentors how they go about supporting that. 

I'll remove it!
